### ISSUE TITLE
Getting INVALID_STATE_ERR with SockJS transformer

### ISSUE THREAD

**Issue by FREEZX** [primus/primus#320]
In a socket benchmark i'm developing, i'm getting the following error when running with sockjs transformer:

```
undefined:1188
        throw new Error('INVALID_STATE_ERR');
        ^
Error: INVALID_STATE_ERR
    at SockJS.send (eval at <anonymous> (/home/freezx/Development/SockBench/node_modules/sockjs-client-node/index.js:15:7), <anonymous>:1188:15)
    at Primus.write (primus.js:1631:24)
    at Primus.emit (events.js:95:17)
    at encoded (primus.js:1066:12)
    at Primus.encoder (primus.js:1660:3)
    at Primus.write [as _write] (primus.js:1060:10)
    at done (primus.js:975:16)
    at transform (primus.js:944:30)
    at Primus.transforms (primus.js:965:4)
    at Primus.write (primus.js:1027:8)
```

You can find the project here:
https://github.com/FREEZX/SockBench

Just do `grunt exec:primus_sockjs` to get the issue.
I don't get the same issue when running the same test with just sockjs without primus at the same version (`grunt exec:sockjs`).

**Comment by 3rd-Eden** [comment]
Closing this, we're using the alpha release of sockjs's new client now. So this should be resolved.

**Comment by 3rd-Eden** [comment]
Errr, seems also to occur on 1.0 of sockjs client. Re-opened.

**Comment by lpinca** [comment]
It took me hours, but this is what I found:

- SockJS sets its `readyState` to `CONNECTING` and tries to open a connection.
- After some time, the connection is opened, SockJS sets its `readyState` to `OPEN` and emits the `open` event.
- The `open` event will schedule an [`incoming::open`](https://github.com/primus/primus/blob/be2db24e3a420267fe675f2147ee4327c069ba8c/transformers/sockjs/client.js#L51) event on primus.
- Now It is possible that our [`connect`](https://github.com/primus/primus/blob/be2db24e3a420267fe675f2147ee4327c069ba8c/primus.js#L997) timer expires right after the SockJS `open` event and this will trigger an [`outoing::reconnect`](https://github.com/primus/primus/blob/be2db24e3a420267fe675f2147ee4327c069ba8c/primus.js#L1010) event.
- The `outgoing::reconnect` event will create a new SockJS instance.
- The deferred `incoming::open` event will now trigger an `open` on primus and if send data, we will get the error because we will use the newly created SockJS instance whose `readyState` is still `CONNECTING`.

Here is a test case:
```js
var Primus = require('primus')
  , http = require('http')
  , primus
  , server;

server = http.createServer();
primus = new Primus(server, { transformer: 'sockjs' });

primus.on('connection', function (spark) {
  spark.pipe(spark);
});

server.listen(3000, function () {
  var socket = primus.Socket('http://localhost:3000', { strategy: false});

  socket.once('end', socket.open);
  socket.once('outgoing::open', function () {
    socket.socket.addEventListener('open', function () {
      socket.end();
    });
  });
  socket.on('open', function () {
    socket.write('data');
  });
});
```
**TL;DR.** This is caused because our events are deferred.

The simplest fix would be to stop deferring events. If I'm not wrong events are deferred because there is a bug in iOS 4.x and 5.x that makes Safari crash. See https://gist.github.com/mloughran/2052006.

Given these [stats](http://david-smith.org/iosversionstats/), I wonder if it makes sense to stop deferring events.

**Comment by sudpaw** [comment]
Hi, i am seeing this error quite often as well. Any news on a fix?

**Comment by lpinca** [comment]
@sudpaw we are discussing if it makes sense to stop deferring events with the risk of having possible crashes in older versions of iOS.
I would go with that but @3rd-Eden, for good reasons, is a bit hesitant.

**Comment by sudpaw** [comment]
Hi again, would a switch to Engine.io instead of SockJs be a possible solution?

**Comment by lpinca** [comment]
@sudpaw the race condition described above can happen with all the transformers.

That said, `engine.io` can be a workaround for this specific issue because it puts the messages in a buffer before sending them.

**Comment by sudpaw** [comment]
Thx :O) I'll give it a go.

**Comment by gatperdut** [comment]
Hi there! Any workaround regarding this woud be appreciated <:^)

**Comment by lpinca** [comment]
Just to make things worse... this race condition leads to data loss when using the `websockets` transformer:
```js
var Primus = require('primus')
  , http = require('http')
  , primus
  , server;

server = http.createServer();
primus = new Primus(server);

primus.on('connection', function (spark) {
  spark.on('data', function (data) {
    console.log('received data: '+ data);
  });
});

server.listen(3000, function () {
  var socket = primus.Socket('http://localhost:3000');

  socket.once('outgoing::open', function () {
    socket.socket.on('open', function () {
      socket.end();
    });
  });
  socket.once('open', function () {
    console.log('writing data');
    socket.write('foo');
  });
});
```
